sql/handler.cc

```diff
@@ -6456,6 +6472,96 @@ void handler::set_lock_type(enum thr_lock_type lock)
   table->reginfo.lock_type= lock;
 }
 
+int handler::ha_check_overlaps(const uchar *old_data, const uchar* new_data)
+{
+  DBUG_ASSERT(new_data);
+  if (!table_share->period.unique_keys)
+    return 0;
+  if (table->versioned() && !table->vers_end_field()->is_max())
+    return 0;
+
+  bool is_update= old_data != NULL;
+  if (!check_overlaps_buffer)
+    check_overlaps_buffer= (uchar*)alloc_root(&table_share->mem_root,
+                                              table_share->max_unique_length
+                                              + table_share->reclength);
+  auto *record_buffer= check_overlaps_buffer + table_share->max_unique_length;
+  auto *handler= this;
+  if (handler->inited != NONE)
+  {
+    if (!check_overlaps_handler)
+    {
+      check_overlaps_handler= clone(table_share->normalized_path.str,
+                                    &table_share->mem_root);
+      int error= -1;
+      if (check_overlaps_handler != NULL)
+        error= check_overlaps_handler->ha_external_lock(table->in_use, F_RDLCK);
+      if (error)
+        return error;
+    }
+    handler= check_overlaps_handler;
+  }
+
+  for (uint key_nr= 0; key_nr < table_share->keys; key_nr++)
+  {
+    const KEY *key_info= table->key_info + key_nr;
+    if (!key_info->without_overlaps)
+      continue;
+
+    key_copy(check_overlaps_buffer, new_data, key_info, 0);
+    key_part_map keypart_map= (1 << key_info->user_defined_key_parts) - 1;
+    for (auto key_func: {HA_READ_KEY_OR_PREV, HA_READ_KEY_OR_NEXT})
```

Comment:
I guess, here should be `{HA_READ_KEY_OR_PREV, HA_READ_AFTER_KEY}`.  There is no need to search for exact key 2 times. Also, exact key match is just for `is_update` case below which should be replaced by (old_data, new_data) check: no need to read index.

Reply:
> There is no need to search for exact key 2 times
That's so, but I believe the node we need would be already cached in most cases, as well as path from the root, so number of IOs should be same (1 IO, actually).

>Also, exact key match is just for is_update case below which should be replaced by (old_data, new_data)~ check: no need to read index.
Those checks are not equal. The important thing is that the fetched key is the same as old_data's key -- that's not exact match; it matches with KEY_OR_PREV, and KEY_OR NEXT

Reply:
changed to `ha_index_read_map` + `ha_index_next`